[DOC] Create Dashboards quickstart (#2409)
opensearch-project/documentation-website: _dashboards/get-started/dql.md

---
layout: default
title: Dashboards Query Language
parent: OpenSearch Dashboards core concepts
nav_order: 40
redirect_from:
  - /dashboards/dql/
---

# Dashboards Query Language

Dashboards Query Language (DQL) is a simple text-based query language for filtering data in OpenSearch Dashboards. Similar to [Query DSL]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/index), DQL uses HTTP request body. For example, to display your site visitor data for a host in the United States, you would enter `geo.dest:US` into the search field.

This documentation uses the Dashboards web log sample data set. See [Quickstart for OpenSearch Dashboards]() to learn how to add sample data to Dashboards.
{: .note}

## Terms query

The most basic query is to specify the search term.

```
host:www.example.com
```

> To access an object's nested field, list the complete path to the field separated by periods. For example, to retrieve the `lat` field in the `coordinates` object:

Reviewer: The end of the sentence could use some clarification. Suggestion: "For example, to retrieve the `lat` field in the `coordinates` object, you can list the following path:"

To avoid confusion, use parentheses to dictate the order in which you want to evaluate. If you want to evaluate `geo.dest:US or response.keyword:200` first, the expression is:

```
(geo.dest:US or response.keyword:200) and host.keyword:www.example.com
```

## Date and range queries

DQL supports numeric inequalities.

```
bytes >= 15 and memory < 15
```

Similarly, you can use the same method to find a date before or after the query. `>` indicates a search for a date after the specified date, and `<` returns dates before.

```
@timestamp > "2020-12-14T09:35:33"
```

## Nested field query

If you have a document with nested fields, you must specify which parts of the document to retrieve.

For example, if you have the following document:

```json
{
  "superheroes":[
    {
      "hero-name": "Superman",
      "real-identity": "Clark Kent",
      "age": 28
    },
    {
      "hero-name": "Batman",
      "real-identity": "Bruce Wayne",
      "age": 26
    },
    {
      "hero-name": "Flash",
      "real-identity": "Barry Allen",
      "age": 28
    },
    {
      "hero-name": "Robin",
      "real-identity": "Dick Grayson",
      "age": 15
    }
  ]
}
```

> The following example shows how to use DQL to retrieve a specific field.

Reviewer 2: These examples are very helpful, but should we also show some example responses after using DQL to retrieve fields?  Like responses when we correctly fetch data and responses when we incorrectly fetch data or empty data.

Author: Done

```
superheroes: {hero-name: Superman}
```

If you want to retrieve multiple objects from your document, specify all the fields you want to retrieve.

```
superheroes: {hero-name: Superman} and superheroes: {hero-name: Batman}
```

The previous boolean and range queries still work, so you can submit a more refined query.

```
superheroes: {hero-name: Superman and age < 50}
```

If your document has an object nested within another object, you can retrieve data by specifying all the levels.

```

> justice-league.superheroes: {hero-name:Superman}

Reviewer 2: Since the original document does not contain `justice-league`, maybe to add the nested field `justice-league` there to avoid confusion?

Author: @abbyhu2000 and @kolchfa-aws Please reveiw the document to confirm accuracy.

```
(geo.dest:US or response.keyword:200) and host.keyword:www.example.com
```

## Querying dates and ranges

DQL supports numeric inequalities, for example, `bytes >= 15 and memory < 15`.

You can use the same method to find a date before or after the date specified in the query. `>` indicates a search for a date after the specified date, and `<` returns dates before the specified date, for example, `@timestamp > "2020-12-14T09:35:33`.

## Querying nested fields

Searching a document with [nested fields]({{site.url}}{{site.baseurl}}/opensearch/supported-field-types/nested/) requires you to specify the full path of the field to be retrieved. In the following example document, the `superheroes` field has nested objects:

```json
{
 "superheroes":[
    {
      "hero-name": "Superman",
      "real-identity": "Clark Kent",
      "age": 28
    },
    {
      "hero-name": "Batman",
      "real-identity": "Bruce Wayne",
      "age": 26
    },
    {
      "hero-name": "Flash",
      "real-identity": "Barry Allen",
      "age": 28
    },
    {
      "hero-name": "Robin",
      "real-identity": "Dick Grayson",
      "age": 15
    }
  ]
}
```
{% include copy.html %}

> To retrieve a specific field using DQL, specify the field, for example:

Author: Change made.

```
superheroes: {hero-name: Superman}
```
{% include copy.html %}

> To retrieve multiple objects from the document, specify all the fields you want to retrieve, for example:

Reviewer 3: ```suggestion
To retrieve documents that match multiple fields, specify all the fields, for example:
```

Author: Change made

```
superheroes: {hero-name: Superman} and superheroes: {hero-name: Batman}
```
{% include copy.html %}

> Continuing with the preceding Boolean and range queries, you can submit a more refined query, for example:

Reviewer 3: ```suggestion
You can combine multiple Boolean and range queries, to create a more refined query, for example:
```

Author: Change made.

```
host.keyword:www.example.com and response.keyword:200
```

The following example shows how to use multiple operators in one query.

```
geo.dest:US or response.keyword:200 and host.keyword:www.example.com
```

Remember that boolean operators follow the logical precedence order of `not`, `and`, and `or`, so if you have an expression like the previous example, `response.keyword:200 and host.keyword:www.example.com` gets evaluated first.

To avoid confusion, use parentheses to dictate the order in which you want to evaluate operands. If you want to evaluate `geo.dest:US or response.keyword:200` first, the expression is:

```
(geo.dest:US or response.keyword:200) and host.keyword:www.example.com
```

## Querying dates and ranges

DQL supports numeric inequalities.

```
bytes >= 15 and memory < 15
```

Similarly, you can use the same method to find a date before or after the query. `>` indicates a search for a date after the specified date, and `<` returns dates before.

```
@timestamp > "2020-12-14T09:35:33"
```

## Querying nested fields

> Searching a document with nested fields requires you to specify which parts of the document to retrieve. In this example, `superheroes` is the nested field type.

Reviewer 3: ```suggestion
Searching a document with [nested fields]({{site.url}}{{site.baseurl}}/opensearch/supported-field-types/nested/) requires you to specify the full path of the field to retrieve. In this example, the `superheroes` field has nested objects:
```

```json
{
 "superheroes":[
    {
      "hero-name": "Superman",
      "real-identity": "Clark Kent",
      "age": 28
    },
    {
      "hero-name": "Batman",
      "real-identity": "Bruce Wayne",
      "age": 26
    },
    {
      "hero-name": "Flash",
      "real-identity": "Barry Allen",
      "age": 28
    },
    {
      "hero-name": "Robin",
      "real-identity": "Dick Grayson",
      "age": 15
    }
  ]
}
```

> To retrieve a specific field using DQL, use the notation:

Reviewer 3: ```suggestion
To retrieve a specific field using DQL, use the following notation:
```

```
superheroes: {hero-name: Superman}
```

To retrieve multiple objects from your document, specify all the fields you want to retrieve, as shown in the following notation:

```
superheroes: {hero-name: Superman} and superheroes: {hero-name: Batman}
```

The previous boolean and range queries still work, so you can submit a more refined query, as shown in the following notation:

```
superheroes: {hero-name: Superman and age < 50}
```

> ## Querying double nested objects 

Reviewer 3: ```suggestion
## Querying doubly nested objects 
```

> If a document has double nested objects (object nested inside another object), retrieve data by specifying all the levels. In the following example document, `superheroes` is double nested.

Reviewer 3: ```suggestion
If a document has doubly nested objects (object nested inside another object), retrieve a field value by specifying the full path to the field. In the following example document, the `superheroes` object is nested inside the `justice-league` object:
```

> 

Reviewer 3: Delete an extra line here

```json

> {

Reviewer 3: This should be :

{
  "justice-league": [
    {
      "superheroes":[
    {
      "hero-name": "Superman",
      "real-identity": "Clark Kent",
      "age": 28
    },
    {
      "hero-name": "Batman",
      "real-identity": "Bruce Wayne",
      "age": 26
    },
    {
      "hero-name": "Flash",
      "real-identity": "Barry Allen",
      "age": 28
    },
    {
      "hero-name": "Robin",
      "real-identity": "Dick Grayson",
      "age": 15
    }
  ]
}
  ]
}

{: .note purple}

## Searching with terms queries

The most basic query specifies the search term:

```
host:www.example.com
```
{% include copy.html %}

To access an object's nested field, list the complete path to the field separated by periods. For example, use the following path to retrieve the `lat` field in the `coordinates` object:

```
coordinates.lat:43.7102
```
{% include copy.html %}

DQL supports leading and trailing wildcards, so you can search for any terms that match your pattern.

```
host.keyword:*.example.com/*
```
{% include copy.html %}

To check whether a field exists or has any data, use a wildcard to see whether Dashboards returns any results:

```
host.keyword:*
```
{% include copy.html %}

## Searching with Boolean queries

To mix and match or combine multiple queries for more refined results, you can use the Boolean operators `and`, `or`, and `not`. DQL is not case sensitive, so `AND` and `and` are the same.

```

> host.keyword:www.example.com and response.keyword:200

Reviewer 4: Reference/introduce the example in the text that precedes it.

Author: Change made.

```
{% include copy.html %}

The following example shows how to use multiple operators in one query:

```
geo.dest:US or response.keyword:200 and host.keyword:www.example.com
```
{% include copy.html %}

Remember that Boolean operators follow the logical precedence order of `not`, `and`, and `or`, so if you have an expression like the one in the preceding example, `response.keyword:200 and host.keyword:www.example.com` is evaluated first.

To avoid confusion, use parentheses to dictate the order in which you want to evaluate operands. If you want to evaluate `geo.dest:US or response.keyword:200` first, use the following expression:

```
(geo.dest:US or response.keyword:200) and host.keyword:www.example.com
```
{% include copy.html %}

## Querying dates and ranges

DQL supports numeric inequalities, for example, `bytes >= 15 and memory < 15`.

You can use the same method to find a date before or after the date specified in the query. `>` indicates a search for a date after the specified date, and `<` returns dates before the specified date, for example, `@timestamp > "2020-12-14T09:35:33`.

## Querying nested fields

> Searching a document with [nested fields]({{site.url}}{{site.baseurl}}/opensearch/supported-field-types/nested/) requires you to specify the full path of the field to be retrieved. In this example, the `superheroes` field has nested objects.

Reviewer 4: ```suggestion
Searching a document with [nested fields]({{site.url}}{{site.baseurl}}/opensearch/supported-field-types/nested/) requires you to specify the full path of the field to be retrieved. In this example, the `superheroes` field has nested objects:
```

Author: Change made

```json
{
 "superheroes":[
    {
      "hero-name": "Superman",
      "real-identity": "Clark Kent",
      "age": 28
    },
    {
      "hero-name": "Batman",
      "real-identity": "Bruce Wayne",
      "age": 26
    },
    {
      "hero-name": "Flash",
      "real-identity": "Barry Allen",
      "age": 28
    },
    {
      "hero-name": "Robin",
      "real-identity": "Dick Grayson",
      "age": 15
    }
  ]
}
```
{% include copy.html %}

To retrieve a specific field using DQL, use the following notation:

```
superheroes: {hero-name: Superman}
```
{% include copy.html %}

To retrieve multiple objects from your document, specify all the fields you want to retrieve, as shown in the following notation:

```
superheroes: {hero-name: Superman} and superheroes: {hero-name: Batman}
```
{% include copy.html %}

> The previous Boolean and range queries still work, so you can submit a more refined query, as shown in the following notation:

Reviewer 4: Same comment re: what we mean by "still work."

Author: Revised. "Continuing with the preceding Boolean and range queries, you can..."

Author: Revised. "Continuing with the preceding Boolean and range queries, you can submit a more refined query, for example:"

```
superheroes: {hero-name: Superman and age < 50}
```
{% include copy.html %}

## Querying doubly nested objects 

> If a document has doubly nested objects (objects nested inside other objects), retrieve a field value by specifying the full path to the field. In the following example document, the `superheroes` object is nested inside the `justice-league` object.

Reviewer 4: ```suggestion
If a document has doubly nested objects (objects nested inside other objects), retrieve a field value by specifying the full path to the field. In the following example document, the `superheroes` object is nested inside the `justice-league` object:
```

Author: Change made

---
layout: default
title: Using Dashboards Query Language
parent: Understanding the core concepts of OpenSearch Dashboards
nav_order: 40
redirect_from:
  - /dashboards/dql/
---

# Using Dashboards Query Language

Dashboards Query Language (DQL) is a simple text-based query language for filtering data in OpenSearch Dashboards. Similar to [Query DSL]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/index), DQL uses HTTP request body. For example, to display your site visitor data for a host in the United States, you would enter `geo.dest:US` into the search field.

> <img src="{{site.url}}{{site.baseurl}}/images/dql-interface.png" alt="DQL toolbar in Dashboard" width="700">

Author: Change made

```
host.keyword:*
```
{% include copy.html %}

## Searching with Boolean queries

To mix and match or combine multiple queries for more refined results, you can use the boolean operators `and`, `or`, and `not`. DQL is not case sensitive, so `AND` and `and` are the same.

```
host.keyword:www.example.com and response.keyword:200
```
{% include copy.html %}

The following example shows how to use multiple operators in one query.

```
geo.dest:US or response.keyword:200 and host.keyword:www.example.com
```
{% include copy.html %}

Remember that boolean operators follow the logical precedence order of `not`, `and`, and `or`, so if you have an expression like the previous example, `response.keyword:200 and host.keyword:www.example.com` gets evaluated first.

To avoid confusion, use parentheses to dictate the order in which you want to evaluate operands. If you want to evaluate `geo.dest:US or response.keyword:200` first, the expression is:

```
(geo.dest:US or response.keyword:200) and host.keyword:www.example.com
```
{% include copy.html %}

## Querying dates and ranges

DQL supports numeric inequalities.

```

> bytes >= 15 and memory < 15

Author: Revised to run example code inline.

```
{% include copy.html %}

Similarly, you can use the same method to find a date before or after the query. `>` indicates a search for a date after the specified date, and `<` returns dates before.

```

> @timestamp > "2020-12-14T09:35:33"

Author: Revised to run example code inline.
